jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -294,6 +297,31 @@ object XGBoost extends Serializable {
     }
   }
 
+  private def repartitionForTrainingGroup(
+      trainingData: RDD[XGBLabeledPoint], nWorkers: Int): RDD[Array[XGBLabeledPoint]] = {
+    val normalGroups = trainingData.mapPartitions(
+      // LabeledPointGroupIterator returns (Boolean, Array[XGBLabeledPoint])
+      new LabeledPointGroupIterator(_)).filter(group => !group._1).map(group => group._2)
+
+    val edgeGroups = trainingData.mapPartitions(
+      new LabeledPointGroupIterator(_)).filter(group => group._1).map(
+        group => (TaskContext.getPartitionId(), group._2))
+    // group chunks from different partitions together by group id in XGBLabeledPoint.
+    val stitchedGroup: RDD[Array[XGBLabeledPoint]] = edgeGroups.groupBy(_._2(0).group).map(r => {
```

Comment:
you may want to have some comments here to explain why groupBy but not aggregateBy